### ISSUE TITLE
reactive forms + ngbDatepicker + ngClass = ExpressionChangedAfterItHasBeenCheckedError

### ISSUE THREAD

**Issue by borriej** [ng-bootstrap/ng-bootstrap#1936]
Im using a large reactive forms set-up, that iterates. A part of the form is:

`<div class="body"
         formArrayName="persons">
        <div class="d-flex person"
             *ngFor="let person of myPersonscontrols; let i=index" [formGroupName]="i">
            <div class="col col-date">
                  <div class="input-group">
                    <input class="form-control"
                          class="form-control"
                           id="date{{i}}"
                           formControlName="date"
                           placeholder="DD/MM/JJJJ"
                           ngbDatepicker
                           #d="ngbDatepicker"
                           [ngClass]="{'is-invalid': person.controls['date'].errors && (person.controls['date'].touched || person.controls['date'].dirty) }">
            </div>`

This causes a javascript error:
`ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'null'. Current value: 'true'.`

which is caused by:
` [ngClass]="{'is-invalid': person.controls['date'].errors && (person.controls['date'].touched || person.controls['date'].dirty) }"`

removing that line solves the error.. 

But why? Is this a issue with reactive forms + ngBootstrap? 

And the values aren't showing using reactive forms.. please add documentation for use without ngModel on typeahead, datepicker etc. 

Thank you!

**Comment by borriej** [comment]
Is it logical that this happens when the date is new Date()?
When i replace it for date = { date.year, date.month, date.day } the error is solved (?)

**Comment by pkozlowski-opensource** [comment]
I'm sorry but there is not enough info here to help you. If you believe that we've got a bug in this library than please provide a minimal reproduce scenario in a plunker: https://github.com/ng-bootstrap/ng-bootstrap/blob/master/README.md#you-think-youve-found-a-bug

If you are unsure if this is a bug or not than please use an appropriate support channel: https://github.com/ng-bootstrap/ng-bootstrap/blob/master/README.md#getting-help

**Comment by pkozlowski-opensource** [comment]
Closing as we haven't heard back from the opener.

**Comment by borriej** [comment]
Sorry, this didn't had anything to do with ng-bootstrap .. ;)

**Comment by BBaysinger** [comment]
I'm having this same issue. If I use date = `{year: 2017, month: 1, day: 11}`, it works, but `{year: 0, month: 0, day: 0}`, or `{year: null, month: null, day: null}` gives `ExpressionChangedAfterItHasBeenCheckedError`.

**Comment by BBaysinger** [comment]
Huh. I figured it out. You have to pass null to the reactive form group if you want the picker to show default. Might be nice if the picker would handle this better with some feedback about an incompatible value being used.

**Comment by fbanyai** [comment]
A bit late to this discussion but... As I just had this issue and by following the tips in this thread I was able to identify new details and solve my problem. So I'd like to share the solution for my case.

The major issues were:

- `NgbDatepicker` has it's own `NgbDateStruct` & `NgbDate` class and it doesn't seem to handle nulls and native Javascript `Date` objects very well. (still investigating)
- I was trusting Typescript to correctly handle a `Date` that was returned by my API (e.g.: 2019-08-13T00:00:00Z). But without any conversion it simply considers it as a **`string`**.

So I ended up using `momentjs` (https://momentjs.com/) to transform a `string` into `NgbDate`.

``` typescript
create(item: Game): FormGroup {
	let released_on: NgbDateStruct;

	if (item.released_on) {
		const date = moment(item.released_on);

		released_on = new NgbDate(date.year(), date.month(), date.day());
	}

	return this.formBuilder.group({
		name: [item.name],
		released_on: [released_on]
	});
}
```
